qiskit/circuit/quantumcircuit.py

```diff
@@ -285,6 +286,45 @@ def __init__(
             raise TypeError("Only a dictionary or None is accepted for circuit metadata")
         self._metadata = metadata
 
+    @staticmethod
+    def from_instructions(
+        instructions: Iterable[
+            CircuitInstruction | tuple[Instruction, Iterable[Qubit], Iterable[Clbit]]
+        ],
+        *,
+        name: Optional[str] = None,
+        global_phase: ParameterValueType = 0,
+        metadata: Optional[dict] = None,
+    ) -> "QuantumCircuit":
+        """Construct a circuit from an iterable of CircuitInstructions.
+
+        Args:
+            instructions: The instructions to add to the circuit.
+            name: The name of the circuit.
+            global_phase: The global phase of the circuit in radians.
+            metadata: Arbitrary key value metadata to associate with the circuit.
+
+        Returns:
+            The quantum circuit.
+        """
+        circuit = QuantumCircuit(name=name, global_phase=global_phase, metadata=metadata)
+        added_qubits = set()
+        added_clbits = set()
+        for instruction in instructions:
+            _, qubits, clbits = instruction
```

Comment:
This does work for `CircuitInstruction`, but it uses a backwards-compatibility shim that at some point we'll be looking to mark as deprecated (though probably not for a couple more Terra versions - I need to update other parts of Qiskit first).  Since we have to do the `isinstance` check in the loop body anyway, perhaps we could raise `instruction` to a `CircuitInstruction` first and then unpack the `qubits` and `clbits` with attribute access?

Reply:
Done.